toolchain/check/import_cpp.cpp

```
@@ -70,13 +68,85 @@ auto ImportCppFiles(
+
+  auto import_loc_id = context.insts().GetLocId(first_import_decl_id);
+
+  auto namespace_inst =
+      SemIR::Namespace{.type_id = namespace_type_id,
+                       .name_scope_id = SemIR::NameScopeId::None,
+                       .import_id = first_import_decl_id};
+
+  auto namespace_inst_and_loc =
+      import_loc_id.is_import_ir_inst_id()
+          ? context.MakeImportedLocAndInst(import_loc_id.import_ir_inst_id(),
+                                           namespace_inst)
+          // TODO: Check that this actually is an `AnyNamespaceId`.
+          : SemIR::LocIdAndInst(Parse::AnyNamespaceId(import_loc_id.node_id()),
+                                namespace_inst);
+
+  auto namespace_id =
+      context.AddPlaceholderInstInNoBlock(namespace_inst_and_loc);
+
+  context.import_ref_ids().push_back(namespace_id);
+  namespace_inst.name_scope_id = context.name_scopes().Add(
+      namespace_id, name_id, SemIR::NameScopeId::Package);
+  context.ReplaceInstBeforeConstantUse(namespace_id, namespace_inst);
+
+  // Note we have to get the parent scope freshly, creating the imported
+  // namespace may invalidate the pointer above.
+  context.name_scopes()
+      .Get(SemIR::NameScopeId::Package)
+      .GetEntry(entry_id)
+      .result = SemIR::ScopeLookupResult::MakeFound(namespace_id,
+                                                    SemIR::AccessKind::Public);
+
+  auto& scope = context.name_scopes().Get(namespace_inst.name_scope_id);
+  scope.set_is_closed_import(true);
+}
+
+auto ImportCppFiles(
+    Context& context, llvm::StringRef importing_file_path,
+    IdentifierId cpp_package_id, SemIR::InstId first_import_decl_id,
+    llvm::ArrayRef<std::pair<llvm::StringRef, SemIRLoc>> imports,
+    llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs) -> void {
+  size_t num_imports = imports.size();
+  if (num_imports == 0) {
+    return;
+  }
+
+  auto ast = GenerateAst(context, importing_file_path, imports, fs);
+  if (!ast) {
+    return;
+  }
```

We should probably do some error recovery here so that Carbon code can be checked as best as possible, even when there are C++ errors. Note this may mean we should try to produce an AST, even when that AST has errors.

Even if an AST cannot be produced, we can still do some error recovery with `scope.set_has_error()`. That should cause name lookups to silently fail, versus the errors that result from declining to add a namespace.

As an example of the behavior goal:

```
import Cpp library "error.h"

// Calling `scope.set_has_error()` should make this silently fail.
// At present, I would expect a `Cpp` name lookup error.
var x: Cpp.Foo;

// This should still diagnose the type mismatch, because it doesn't touch `Cpp`.
var x: i32 = 1.0;
```

So for example here:

```
auto ast = GenerateAst(...);
auto name_scope_id = AddNamespace(...);
if (!ast) {
  context.name_scopes().Get(name_scope_id).set_has_error();
}
```

---

ok, I now always returns the pointer to the AST and mark the scope with error if it's nullptr.
We can still have errors when generating the AST, but if it's created it would not be marked with errors.
I'm not sure whether we want to mark the scope with `has_error` if Clang triggered errors but some AST was created, so I took the conservative (and simpler) approach.